go/vt/topo/etcd2topo/server_test.go

```diff
@@ -89,6 +95,122 @@ func startEtcd(t *testing.T) (*exec.Cmd, string, string) {
 	return cmd, dataDir, clientAddr
 }
 
+// startEtcdWithTLS starts an etcd subprocess with TLS setup, and waits for it to be ready.
+func startEtcdWithTLS(t *testing.T) (*exec.Cmd, string, string) {
+	// Create a temporary directory.
+	dataDir, err := ioutil.TempDir("", "etcd")
+	if err != nil {
+		t.Fatalf("cannot create tempdir: %v", err)
+	}
+
+	// Get our two ports to listen to.
+	port := testfiles.GoVtTopoEtcd2topoPort
+	name := "vitess_unit_test"
+	clientAddr := fmt.Sprintf("http://localhost:%v", port)
+	peerAddr := fmt.Sprintf("http://localhost:%v", port+1)
+	initialCluster := fmt.Sprintf("%v=%v", name, peerAddr)
+
+	// Generate cert and key in /tmp
+	err = exec.Command("openssl", "req", "-new", "-newkey", "rsa:4096", "-days", "365", "-nodes", "-x509",
+		"-subj", "/C=US/ST=Denial/L=Springfield/O=Dis/CN=www.example.com",
+		"-keyout", testKeyPath, "-out", testCertPath).Run()
+	if err != nil {
+		t.Fatalf("unable to generate test key and cert using openssl: %v", err)
+	}
+
+	cmd := exec.Command("etcd",
+		"-name", name,
+		"-advertise-client-urls", clientAddr,
+		"-initial-advertise-peer-urls", peerAddr,
+		"-listen-client-urls", clientAddr,
+		"-listen-peer-urls", peerAddr,
+		"-initial-cluster", initialCluster,
+		"-cert-file", testCertPath,
+		"-key-file", testKeyPath,
+		"-data-dir", dataDir)
+	err = cmd.Start()
+	if err != nil {
+		t.Fatalf("failed to start etcd: %v", err)
+	}
+
+	// Safe now to build up TLS info.
+	tlsInfo := transport.TLSInfo{
+		CertFile:   testCertPath,
+		KeyFile:    testKeyPath,
+		TrustedCAFile: *caPath,
+	}
+
+	tlsConfig, err := tlsInfo.ClientConfig()
+
+	// Create a client to connect to the created etcd.
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{clientAddr},
+		TLS: tlsConfig,
+		DialTimeout: 5 * time.Second,
+	})
+	if err != nil {
+		t.Fatalf("newCellClient(%v) failed: %v", clientAddr, err)
+	}
+
+	// Wait until we can list "/", or timeout.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	start := time.Now()
+	for {
+		if _, err := cli.Get(ctx, "/"); err == nil {
+			break
+		}
+		if time.Since(start) > 10*time.Second {
+			t.Fatalf("Failed to start etcd daemon in time")
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	return cmd, dataDir, clientAddr
+}
+
+func TestEtcd2TLS(t *testing.T) {
+	// Start a single etcd in the background.
+	cmd, dataDir, clientAddr := startEtcdWithTLS(t)
+	defer func() {
+		cmd.Process.Kill()
+		cmd.Wait()
+		os.RemoveAll(dataDir)
+	}()
+
+	testIndex := 0
+	testRoot := fmt.Sprintf("/test-%v", testIndex)
+	// Create the server on the new root.
+	server, err := NewServerWithOpts(clientAddr, testRoot, &testCertPath, &testKeyPath, &testCertPath)
+	if err != nil {
+		t.Fatalf("NewServerWithOpts failed: %v", err)
+	}
+
+	client := server.cli
+	// If TLS is not setup correctly, this will fail
+	_, err = client.Dial(clientAddr)
```

Comment:
You shouldn't have to Dial yourself. We should just test that we can call `Create()` and `Get()` on `server` so we're testing it the way a caller would use the server, instead of reaching in to directly use the private `cli` member.

Reply:
Done.